decred/decred/dcr/wire/netaddress.py

```diff
@@ -0,0 +1,222 @@
+"""
+Copyright (c) 2020, the Decred developers
+See LICENSE for details
+"""
+
+import socket
+import time
+
+from decred import DecredError
+from decred.util.encode import ByteArray
+
+
+MaxNetAddressPayload = 30
+
+# Prefix for a IPv4 adderess encoded as 16 bytes.
+ipv4to16prefix = ByteArray(0xFFFF, length=12)
+
+
+class NetAddress:
+    """
+	NetAddress defines information about a peer on the network including the time
+	it was last seen, the services it supports, its IP address, and port.
+	"""
+
+    def __init__(self, ip, port, services, stamp):
+        """
+		Args:
+			ip (str or bytes-like): The peer's IP address.
+			port (int): Port the peer is using.  This is encoded in big endian
+				on the wire which differs from most everything else.
+			services (int): Bitfield which identifies the services supported by
+				the peer.
+			stamp (int): The last time the peer was seen. This is,
+				unfortunately, encoded as an int on the wire and therefore is
+				limited to 2106. This field is not present in the Decred version
+				message (MsgVersion) nor was it added until protocol
+				version >= NetAddressTimeVersion.
+		"""
+        self.timestamp = stamp
+        self.services = services
+
+        # If the IP is a string, parse it to bytes.
+        if isinstance(ip, str):
+            ip = decodeStringIP(ip)
+        self.ip = ip
+
+        self.port = port
+
+    def hasService(self, service):
+        """
+		Whether the specified service is supported by the address.
+
+		Args:
+			service (int): Bitfield which identifies the service(s) to check.
+
+		Returns:
+			bool: True if this peer offers the service(s).
+		"""
+        return self.services & service == service
+
+    def addService(self, service):
+        """
+		Adds service(s) as a supported service.
+
+		Args:
+			service (int): Bitfield which identifies the service(s) to add.
+		"""
+        self.services |= service
+
+
+def newNetAddressIPPort(ip, port, services):
+    """
+	A new NetAddress using the provided IP, port, and supported services with
+	defaults for the remaining fields.
+
+	Args:
+		ip (str or bytes-like): The peer's IP address.
+		port (int): The peer's port.
+		services (int): Bitfield which identifies the services supported by
+			the peer.
+
+	Returns:
+		NetAddress: The peer's NetAddress.
+	"""
+    return newNetAddressTimestamp(int(time.time()), services, ip, port)
+
+
+def newNetAddressTimestamp(stamp, services, ip, port):
+    """
+	A new NetAddress using the provided timestamp, IP, port, and supported
+	services.
+
+	Args:
+		stamp (int): The last time the peer was seen.
+		ip (str or bytes-like): The peer's IP address.
+		port (int): The peer's port.
+		services (int): Bitfield which identifies the services supported by
+			the peer.
+
+	Returns:
+		NetAddress: The peer's NetAddress.
+	"""
+    return NetAddress(ip, port, services, stamp)
+
+
+def newNetAddress(tcpAddr, services):
```

Comment:
This should be called `newNetAddressTcpAddr` or similar, to emphasize that it's not the complete factory function.